src/storage/FileResourceStore.ts

```diff
@@ -0,0 +1,458 @@
+        if (!stats.isDirectory()) {
+          reject(new MethodNotAllowedHttpError('The given path is not a valid container.'));
+        } else {
+          if (metadata) {
+            this.createMetadata(`${path}${resourceName}.metadata`, metadata, reject);
+          }
+          const writeStream = createWriteStream(path + resourceName);
+
+          data.on('data', (chunk): any => writeStream.write(chunk));
+          data.on('error', reject);
+          data.on('end', (): any => writeStream.end());
+
+          writeStream.on('error', reject);
+          writeStream.on('finish', (): any => resolve({ path: this.mapFilepathToUrl(path + resourceName) }));
+        }
+      } catch (error) {
+        reject(new MethodNotAllowedHttpError());
+      }
+    });
+  }
+
+  /**
+   * Create a directory to represent a container.
+   * @param path - The path to the parent directory in which the new directory should be created.
+   * @param _containerName - The name of the directory to be created.
+   * @param allowRecursiveCreation - Whether necessary but not existing intermediate containers may be created.
+   * @param metadata - Optional metadata that will be stored at `path/_containerName/.metadata` if set.
+   *
+   * @returns Promise of the identifier of the newly created container.
+   */
+  private async createContainer(path: string, _containerName: string,
+    allowRecursiveCreation: boolean, metadata?: Readable): Promise<ResourceIdentifier> {
+    const containerName = ensureTrailingSlash(_containerName);
+    return new Promise(async(resolve, reject): Promise<any> => {
+      try {
+        if (allowRecursiveCreation) {
+          await fsPromises.mkdir(path + containerName, { recursive: true });
+          if (metadata) {
+            this.createMetadata(`${path}${containerName}.metadata`, metadata, reject);
+          }
+          resolve({ path: this.mapFilepathToUrl(path + containerName) });
+        } else {
+          const stats = await fsPromises.lstat(path);
+          if (!stats.isDirectory()) {
+            reject(new MethodNotAllowedHttpError('The given path is not a valid container.'));
+          } else {
+            await fsPromises.mkdir(path + containerName, { recursive: false });
+            if (metadata) {
+              this.createMetadata(`${path}${containerName}.metadata`, metadata, reject);
+            }
+            resolve({ path: this.mapFilepathToUrl(path + containerName) });
+          }
+        }
+      } catch (error) {
+        reject(new MethodNotAllowedHttpError());
+      }
+    });
+  }
+
+  /**
+   * Helper function without extra validation checking to create a metadata file.
+   * @param path - The filepath of the file to be created.
+   * @param metadata - The data to be put in the file.
+   * @param reject - The function to call when an error occur.
+   */
+  private createMetadata(path: string, metadata: Readable, reject: (reason?: any) => void): void {
```

Comment:
This can be used cleaner in other functions if this returns a `Promise<void>` instead of taking a `reject` function as input.